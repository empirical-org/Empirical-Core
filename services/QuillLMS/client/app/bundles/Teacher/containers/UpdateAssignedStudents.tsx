import * as React from 'react';

import { requestGet, requestPut, requestPost, } from '../../../modules/request';
import { Spinner, } from '../../Shared/index';
import Student from '../components/update_assigned_students/student'
import UnassignWarningModal from '../components/update_assigned_students/unassign_warning_modal'
import { unorderedArraysAreEqual, } from '../../../modules/unorderedArraysAreEqual'

const UpdateAssignedStudents = ({ match, unassignWarningHidden, skipLoading, passedOriginalClassrooms, passedClassroomsForComparison, passedAssignmentData, passedUnitName, }) => {
  const [loading, setLoading] = React.useState(skipLoading ? false : true)
  const [originalClassrooms, setOriginalClassrooms] = React.useState(null || passedOriginalClassrooms)
  const [classroomsForComparison, setClassroomsForComparison] = React.useState(null || passedClassroomsForComparison)
  const [assignmentData, setAssignmentData] = React.useState(null || passedAssignmentData)
  const [unitName, setUnitName] = React.useState(null || passedUnitName)
  const [showUnassignWarningModal, setShowUnassignWarningModal] = React.useState(false)
  const [hideWarningModalInFuture, setHideWarningModalInFuture] = React.useState(false)

  React.useEffect(() => {
    getClassroomsAndStudentsData(true)
  }, [])

  function getClassroomsAndStudentsData(firstTime=false) {
    requestGet(`/teachers/units/${match.params.unitId}/classrooms_with_students_and_classroom_units`, (body) => {
      const newAssignmentData = body.classrooms.map(c => ({
        id: c.id,
        student_ids: c.classroom_unit?.assigned_student_ids || [],
        assign_on_join: c.classroom_unit?.assign_on_join || false
      }))

      if (firstTime) {
        setOriginalClassrooms(body.classrooms)
      }

      setAssignmentData(newAssignmentData)
      setClassroomsForComparison(body.classrooms)
      setUnitName(body.unit_name)
      setLoading(false)
    })
  };

  function removedStudentIds() {
    const ids = []

    assignmentData.forEach(assignment => {
      const classroom = classroomsForComparison.find(c => c.id === assignment.id)
      if (!classroom.classroom_unit) { return }

      ids.push(classroom.classroom_unit.assigned_student_ids.filter(id => !assignment.student_ids.includes(id)))
    })

    return ids.flat()
  }

  function handleClickUpdate() {
    if (removedStudentIds().length && !unassignWarningHidden) {
      setShowUnassignWarningModal(true)
    } else {
      updateAssignedStudents()
    }
  }

  function updateAssignedStudents() {
    const filteredAssignmentData = assignmentData.filter(assignment => {
      const existingClassroom = classroomsForComparison.find(c => c.id === assignment.id)
      return existingClassroom.classroom_unit || assignment.student_ids.length
    })
    setLoading(true)
    requestPut(`/teachers/units/${match.params.unitId}/update_classroom_unit_assigned_students`, { unit: { classrooms: filteredAssignmentData }}, (body) => {
      getClassroomsAndStudentsData()
    })
  }

  function assignedStudentsHaveChanged() {
    return assignmentData.some(assignment => {
      const classroom = classroomsForComparison.find(c => c.id === assignment.id)
      return !unorderedArraysAreEqual(assignment.student_ids, (classroom.classroom_unit?.assigned_student_ids || []))
    })
  }

  function toggleStudentSelection(studentId, classroomId) {
    const newAssignmentData = [...assignmentData]
    const newAssignment = newAssignmentData.find(a => a.id === classroomId)
    const classroom = classroomsForComparison.find(c => c.id === classroomId)
    newAssignment.student_ids = newAssignment.student_ids.includes(studentId) ? newAssignment.student_ids.filter(id => id !== studentId) : [...newAssignment.student_ids, studentId]
    newAssignment.assign_on_join = newAssignment.student_ids.length === classroom.students.length

    setAssignmentData(newAssignmentData)
  }

  function revertUnassignment(studentId, classroomUnitId) {
    requestPut(`/teachers/units/${match.params.unitId}/restore_classroom_unit_assignment_for_one_student`, { classroom_unit_id: classroomUnitId, student_id: studentId }, (body) => {
      getClassroomsAndStudentsData()
    })
  }

  function closeWarningModal() {
    setShowUnassignWarningModal(false)
  }

  function sendRequestToHideWarningModalInFuture() {
    requestPost(
      `${process.env.DEFAULT_URL}/milestones/complete_dismiss_unassign_warning_modal`,
      {},
      (body) => {}
    )
  }

  function onClickWarningModalUpdate() {
    if (hideWarningModalInFuture) {
      sendRequestToHideWarningModalInFuture()
    }
    updateAssignedStudents()
    closeWarningModal()
  }

  function toggleHideWarningModalInFuture() {
    setHideWarningModalInFuture(!hideWarningModalInFuture)
  }

  if (loading) {
    return <Spinner />
  }

  const classrooms = classroomsForComparison.map(classroom => {
    const originalClassroom = originalClassrooms.find(c => c.id === classroom.id)
    const assignment = assignmentData.find(a => a.id === classroom.id)
    const numberOfStudentsToAssign = assignment.student_ids.length

    const unselectAll = () => {
      const newAssignmentData = [...assignmentData]
      const newAssignment = newAssignmentData.find(a => a.id === classroom.id)
      newAssignment.student_ids = []
      newAssignment.assign_on_join = false
      setAssignmentData(newAssignmentData)
    }

    const selectAll = () => {
      const newAssignmentData = [...assignmentData]
      const newAssignment = newAssignmentData.find(a => a.id === classroom.id)
      newAssignment.student_ids = classroom.students.map(s => s.id)
      newAssignment.assign_on_join = true
      setAssignmentData(newAssignmentData)
    }

    const students = classroom.students.map(s => (
      <Student
        assignment={assignment}
        classroomUnit={classroom.classroom_unit}
        key={s.id}
        originalClassroomUnit={originalClassroom.classroom_unit}
        revertUnassignment={revertUnassignment}
        student={s}
        toggleStudentSelection={toggleStudentSelection}
      />
    ))

    return (
      <section className="classroom-section" key={classroom.id}>
        <div className="classroom-header">
          <div>
            <h3>{classroom.name}</h3>
            <span>{numberOfStudentsToAssign}/{classroom.students.length} student{numberOfStudentsToAssign === 1 ? '' : 's'} will be assigned</span>
          </div>
          <div className="buttons">
            <button className="quill-button secondary outlined fun focus-on-dark" onClick={unselectAll} type="button">Unselect all</button>
            <button className="quill-button contained primary fun focus-on-dark" onClick={selectAll} type="button">Select all</button>
          </div>
        </div>
        <div className="students">
          {students}
        </div>
      </section>
    )
  })

  return (
    <section className="update-assigned-students-container white-background-accommodate-footer">
      <div className="container">
        {showUnassignWarningModal && (
          <UnassignWarningModal
            closeModal={closeWarningModal}
            handleClickUpdate={onClickWarningModalUpdate}
            hideWarningModalInFuture={hideWarningModalInFuture}
            removedStudentCount={removedStudentIds().length}
            toggleCheckbox={toggleHideWarningModalInFuture}
          />
        )}
        <div className="container">
          <section className="header">
            <div>
              <h1>Update students assigned to activity pack</h1>
              <h2>{unitName}</h2>
            </div>
            <div className="buttons">
              <a className={`quill-button outlined secondary medium focus-on-light ${assignedStudentsHaveChanged() ? '' : 'disabled'}`} href="/teachers/classrooms/activity_planner">Cancel</a>
              <button className={`quill-button contained primary medium focus-on-light ${assignedStudentsHaveChanged() ? '' : 'disabled'}`} disabled={!assignedStudentsHaveChanged()} onClick={handleClickUpdate} type="button">{assignedStudentsHaveChanged() ? 'Update students assigned to pack' : 'Edit students before saving'}</button>
            </div>
          </section>
          {classrooms}
        </div>
      </div>
    </section>
  )
}

export default UpdateAssignedStudents
