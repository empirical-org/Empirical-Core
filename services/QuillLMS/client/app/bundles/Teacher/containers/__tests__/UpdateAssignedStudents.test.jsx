import {
  mount,
  shallow
} from 'enzyme';
import _ from 'lodash';
import React from 'react';

import UpdateAssignedStudents from '../UpdateAssignedStudents';
import { classroomsWithStudentsContainerData, } from './test_data';
