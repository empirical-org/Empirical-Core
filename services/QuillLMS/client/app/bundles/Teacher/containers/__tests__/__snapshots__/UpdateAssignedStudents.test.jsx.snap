// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAssignedStudents will render  1`] = `
<UpdateAssignedStudents
  match={
    Object {
      "params": Object {
        "unitId": 912,
      },
    }
  }
  passedAssignmentData={
    Array [
      Object {
        "assign_on_join": false,
        "id": 173,
        "student_ids": Array [
          1032,
          1036,
        ],
      },
    ]
  }
  passedClassroomsForComparison={
    Array [
      Object {
        "classroom_unit": Object {
          "assign_on_join": false,
          "assigned_student_ids": Array [
            1032,
            1034,
          ],
          "id": 1067,
        },
        "clever_id": null,
        "code": "demo-37",
        "created_at": "2023-04-25T15:53:50.233Z",
        "google_classroom_id": null,
        "grade": "9",
        "grade_level": null,
        "id": 173,
        "name": "Quill Classroom",
        "students": Array [
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:52.582Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1032,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Ken Liu",
            "password_digest": "$2a$10$8Z36DFcncF.9BmXT78QZROl1lWtr.5aogEmzMUDrhLHzWrD1NeIxG",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:52.582Z",
            "username": "ken.liu.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:53.083Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1035,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Tahereh Mafi",
            "password_digest": "$2a$10$ti487x9/9Tn88SklNjQ5zuhvCuf5aV3QTfe55vAoCkQ9sYUp6Nz.C",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:53.083Z",
            "username": "tahereh.mafi.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:52.907Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1033,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Jason Reynolds",
            "password_digest": "$2a$10$98a7KLY.t7c6FpRA1RUsiOEpuuXCDKS4PXSaTN1fPesxIbQuU6iZK",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:52.907Z",
            "username": "jason.reynolds.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:52.992Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1034,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Nic Stone",
            "password_digest": "$2a$10$y0pCHpdp4W6CszSUGD/MBeoUxPcj6wrzgL7nce/n3gi65cpszyoG6",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:52.992Z",
            "username": "nic.stone.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:53.181Z",
            "email": "angie_thomas_demo@quill.org",
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1036,
            "ip_address": "127.0.0.1",
            "last_active": "2023-05-01T13:39:37.000Z",
            "last_sign_in": "2023-05-01T13:36:42.845Z",
            "name": "Angie Thomas",
            "password_digest": "$2a$10$elzHkfkS1qXiq40ohtYUTezooxkdS9PxqmFrKon7vU87P1A3KWNXC",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-05-01T13:39:38.357Z",
            "username": "angie.thomas.173@demo-teacher",
          },
        ],
        "synced_name": null,
        "teacher_id": null,
        "updated_at": "2023-07-06T18:38:38.205Z",
        "visible": true,
      },
    ]
  }
  passedOriginalClassrooms={
    Array [
      Object {
        "classroom_unit": Object {
          "assign_on_join": false,
          "assigned_student_ids": Array [
            1032,
            1034,
          ],
          "id": 1067,
        },
        "clever_id": null,
        "code": "demo-37",
        "created_at": "2023-04-25T15:53:50.233Z",
        "google_classroom_id": null,
        "grade": "9",
        "grade_level": null,
        "id": 173,
        "name": "Quill Classroom",
        "students": Array [
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:52.582Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1032,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Ken Liu",
            "password_digest": "$2a$10$8Z36DFcncF.9BmXT78QZROl1lWtr.5aogEmzMUDrhLHzWrD1NeIxG",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:52.582Z",
            "username": "ken.liu.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:53.083Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1035,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Tahereh Mafi",
            "password_digest": "$2a$10$ti487x9/9Tn88SklNjQ5zuhvCuf5aV3QTfe55vAoCkQ9sYUp6Nz.C",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:53.083Z",
            "username": "tahereh.mafi.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:52.907Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1033,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Jason Reynolds",
            "password_digest": "$2a$10$98a7KLY.t7c6FpRA1RUsiOEpuuXCDKS4PXSaTN1fPesxIbQuU6iZK",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:52.907Z",
            "username": "jason.reynolds.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:52.992Z",
            "email": null,
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1034,
            "ip_address": null,
            "last_active": null,
            "last_sign_in": null,
            "name": "Nic Stone",
            "password_digest": "$2a$10$y0pCHpdp4W6CszSUGD/MBeoUxPcj6wrzgL7nce/n3gi65cpszyoG6",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-04-25T15:53:52.992Z",
            "username": "nic.stone.173@demo-teacher",
          },
          Object {
            "account_type": "unknown",
            "active": false,
            "classcode": null,
            "clever_id": null,
            "created_at": "2023-04-25T15:53:53.181Z",
            "email": "angie_thomas_demo@quill.org",
            "flags": Array [],
            "flagset": "production",
            "google_id": null,
            "id": 1036,
            "ip_address": "127.0.0.1",
            "last_active": "2023-05-01T13:39:37.000Z",
            "last_sign_in": "2023-05-01T13:36:42.845Z",
            "name": "Angie Thomas",
            "password_digest": "$2a$10$elzHkfkS1qXiq40ohtYUTezooxkdS9PxqmFrKon7vU87P1A3KWNXC",
            "role": "student",
            "send_newsletter": false,
            "signed_up_with_google": false,
            "stripe_customer_id": null,
            "time_zone": null,
            "title": null,
            "token": null,
            "updated_at": "2023-05-01T13:39:38.357Z",
            "username": "angie.thomas.173@demo-teacher",
          },
        ],
        "synced_name": null,
        "teacher_id": null,
        "updated_at": "2023-07-06T18:38:38.205Z",
        "visible": true,
      },
    ]
  }
  passedUnitName="Quill Activity Pack"
  skipLoading={true}
  unassignWarningHidden={false}
>
  <section
    className="update-assigned-students-container white-background-accommodate-footer"
  >
    <div
      className="container"
    >
      <div
        className="container"
      >
        <section
          className="header"
        >
          <div>
            <h1>
              Update students assigned to activity pack
            </h1>
            <h2>
              Quill Activity Pack
            </h2>
          </div>
          <div
            className="buttons"
          >
            <a
              className="quill-button outlined secondary medium focus-on-light "
              href="/teachers/classrooms/activity_planner"
            >
              Cancel
            </a>
            <button
              className="quill-button contained primary medium focus-on-light "
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Update students assigned to pack
            </button>
          </div>
        </section>
        <section
          className="classroom-section"
          key="173"
        >
          <div
            className="classroom-header"
          >
            <div>
              <h3>
                Quill Classroom
              </h3>
              <span>
                2
                /
                5
                 student
                s
                 will be assigned
              </span>
            </div>
            <div
              className="buttons"
            >
              <button
                className="quill-button secondary outlined fun focus-on-dark"
                onClick={[Function]}
                type="button"
              >
                Unselect all
              </button>
              <button
                className="quill-button contained primary fun focus-on-dark"
                onClick={[Function]}
                type="button"
              >
                Select all
              </button>
            </div>
          </div>
          <div
            className="students"
          >
            <Student
              assignment={
                Object {
                  "assign_on_join": false,
                  "id": 173,
                  "student_ids": Array [
                    1032,
                    1036,
                  ],
                }
              }
              classroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              key="1032"
              originalClassroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              revertUnassignment={[Function]}
              student={
                Object {
                  "account_type": "unknown",
                  "active": false,
                  "classcode": null,
                  "clever_id": null,
                  "created_at": "2023-04-25T15:53:52.582Z",
                  "email": null,
                  "flags": Array [],
                  "flagset": "production",
                  "google_id": null,
                  "id": 1032,
                  "ip_address": null,
                  "last_active": null,
                  "last_sign_in": null,
                  "name": "Ken Liu",
                  "password_digest": "$2a$10$8Z36DFcncF.9BmXT78QZROl1lWtr.5aogEmzMUDrhLHzWrD1NeIxG",
                  "role": "student",
                  "send_newsletter": false,
                  "signed_up_with_google": false,
                  "stripe_customer_id": null,
                  "time_zone": null,
                  "title": null,
                  "token": null,
                  "updated_at": "2023-04-25T15:53:52.582Z",
                  "username": "ken.liu.173@demo-teacher",
                }
              }
              toggleStudentSelection={[Function]}
            >
              <div
                className="student already-assigned"
              >
                <div>
                  <button
                    aria-label="Checked checkbox"
                    className="quill-checkbox selected"
                    onClick={[Function]}
                    type="button"
                  >
                    <img
                      alt="Checked checkbox"
                      src="undefined/images/shared/check-small-white.svg"
                    />
                  </button>
                  <span>
                    Ken Liu
                  </span>
                </div>
                <span>
                  Pack 
                  <em>
                    assigned
                  </em>
                </span>
              </div>
            </Student>
            <Student
              assignment={
                Object {
                  "assign_on_join": false,
                  "id": 173,
                  "student_ids": Array [
                    1032,
                    1036,
                  ],
                }
              }
              classroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              key="1035"
              originalClassroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              revertUnassignment={[Function]}
              student={
                Object {
                  "account_type": "unknown",
                  "active": false,
                  "classcode": null,
                  "clever_id": null,
                  "created_at": "2023-04-25T15:53:53.083Z",
                  "email": null,
                  "flags": Array [],
                  "flagset": "production",
                  "google_id": null,
                  "id": 1035,
                  "ip_address": null,
                  "last_active": null,
                  "last_sign_in": null,
                  "name": "Tahereh Mafi",
                  "password_digest": "$2a$10$ti487x9/9Tn88SklNjQ5zuhvCuf5aV3QTfe55vAoCkQ9sYUp6Nz.C",
                  "role": "student",
                  "send_newsletter": false,
                  "signed_up_with_google": false,
                  "stripe_customer_id": null,
                  "time_zone": null,
                  "title": null,
                  "token": null,
                  "updated_at": "2023-04-25T15:53:53.083Z",
                  "username": "tahereh.mafi.173@demo-teacher",
                }
              }
              toggleStudentSelection={[Function]}
            >
              <div
                className="student "
              >
                <div>
                  <button
                    aria-label="Unchecked checkbox"
                    className="quill-checkbox unselected"
                    onClick={[Function]}
                    type="button"
                  />
                  <span>
                    Tahereh Mafi
                  </span>
                </div>
              </div>
            </Student>
            <Student
              assignment={
                Object {
                  "assign_on_join": false,
                  "id": 173,
                  "student_ids": Array [
                    1032,
                    1036,
                  ],
                }
              }
              classroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              key="1033"
              originalClassroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              revertUnassignment={[Function]}
              student={
                Object {
                  "account_type": "unknown",
                  "active": false,
                  "classcode": null,
                  "clever_id": null,
                  "created_at": "2023-04-25T15:53:52.907Z",
                  "email": null,
                  "flags": Array [],
                  "flagset": "production",
                  "google_id": null,
                  "id": 1033,
                  "ip_address": null,
                  "last_active": null,
                  "last_sign_in": null,
                  "name": "Jason Reynolds",
                  "password_digest": "$2a$10$98a7KLY.t7c6FpRA1RUsiOEpuuXCDKS4PXSaTN1fPesxIbQuU6iZK",
                  "role": "student",
                  "send_newsletter": false,
                  "signed_up_with_google": false,
                  "stripe_customer_id": null,
                  "time_zone": null,
                  "title": null,
                  "token": null,
                  "updated_at": "2023-04-25T15:53:52.907Z",
                  "username": "jason.reynolds.173@demo-teacher",
                }
              }
              toggleStudentSelection={[Function]}
            >
              <div
                className="student "
              >
                <div>
                  <button
                    aria-label="Unchecked checkbox"
                    className="quill-checkbox unselected"
                    onClick={[Function]}
                    type="button"
                  />
                  <span>
                    Jason Reynolds
                  </span>
                </div>
              </div>
            </Student>
            <Student
              assignment={
                Object {
                  "assign_on_join": false,
                  "id": 173,
                  "student_ids": Array [
                    1032,
                    1036,
                  ],
                }
              }
              classroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              key="1034"
              originalClassroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              revertUnassignment={[Function]}
              student={
                Object {
                  "account_type": "unknown",
                  "active": false,
                  "classcode": null,
                  "clever_id": null,
                  "created_at": "2023-04-25T15:53:52.992Z",
                  "email": null,
                  "flags": Array [],
                  "flagset": "production",
                  "google_id": null,
                  "id": 1034,
                  "ip_address": null,
                  "last_active": null,
                  "last_sign_in": null,
                  "name": "Nic Stone",
                  "password_digest": "$2a$10$y0pCHpdp4W6CszSUGD/MBeoUxPcj6wrzgL7nce/n3gi65cpszyoG6",
                  "role": "student",
                  "send_newsletter": false,
                  "signed_up_with_google": false,
                  "stripe_customer_id": null,
                  "time_zone": null,
                  "title": null,
                  "token": null,
                  "updated_at": "2023-04-25T15:53:52.992Z",
                  "username": "nic.stone.173@demo-teacher",
                }
              }
              toggleStudentSelection={[Function]}
            >
              <div
                className="student will-be-removed"
              >
                <div>
                  <button
                    aria-label="Unchecked checkbox"
                    className="quill-checkbox unselected"
                    onClick={[Function]}
                    type="button"
                  />
                  <span>
                    Nic Stone
                  </span>
                </div>
                <span>
                  Pack will be 
                  <em>
                    removed
                  </em>
                </span>
              </div>
            </Student>
            <Student
              assignment={
                Object {
                  "assign_on_join": false,
                  "id": 173,
                  "student_ids": Array [
                    1032,
                    1036,
                  ],
                }
              }
              classroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              key="1036"
              originalClassroomUnit={
                Object {
                  "assign_on_join": false,
                  "assigned_student_ids": Array [
                    1032,
                    1034,
                  ],
                  "id": 1067,
                }
              }
              revertUnassignment={[Function]}
              student={
                Object {
                  "account_type": "unknown",
                  "active": false,
                  "classcode": null,
                  "clever_id": null,
                  "created_at": "2023-04-25T15:53:53.181Z",
                  "email": "angie_thomas_demo@quill.org",
                  "flags": Array [],
                  "flagset": "production",
                  "google_id": null,
                  "id": 1036,
                  "ip_address": "127.0.0.1",
                  "last_active": "2023-05-01T13:39:37.000Z",
                  "last_sign_in": "2023-05-01T13:36:42.845Z",
                  "name": "Angie Thomas",
                  "password_digest": "$2a$10$elzHkfkS1qXiq40ohtYUTezooxkdS9PxqmFrKon7vU87P1A3KWNXC",
                  "role": "student",
                  "send_newsletter": false,
                  "signed_up_with_google": false,
                  "stripe_customer_id": null,
                  "time_zone": null,
                  "title": null,
                  "token": null,
                  "updated_at": "2023-05-01T13:39:38.357Z",
                  "username": "angie.thomas.173@demo-teacher",
                }
              }
              toggleStudentSelection={[Function]}
            >
              <div
                className="student will-be-assigned"
              >
                <div>
                  <button
                    aria-label="Checked checkbox"
                    className="quill-checkbox selected"
                    onClick={[Function]}
                    type="button"
                  >
                    <img
                      alt="Checked checkbox"
                      src="undefined/images/shared/check-small-white.svg"
                    />
                  </button>
                  <span>
                    Angie Thomas
                  </span>
                </div>
                <span>
                  Pack will be 
                  <em>
                    assigned
                  </em>
                </span>
              </div>
            </Student>
          </div>
        </section>
      </div>
    </div>
  </section>
</UpdateAssignedStudents>
`;
